include/core/mir/geometry/dimensions_generic.h

```diff
@@ -0,0 +1,290 @@
+/*
+ * Copyright © 2020 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored by: William Wold <william.wold@canonical.com>
+ */
+
+#ifndef MIR_GEOMETRY_DIMENSIONS_GENERIC_H_
+#define MIR_GEOMETRY_DIMENSIONS_GENERIC_H_
+
+#include <iosfwd>
+#include <type_traits>
+
+namespace mir
+{
+/// Basic geometry types. Types for dimensions, displacements, etc.
+/// and the operations that they support.
+namespace geometry
+{
+
+/// These tag types determine what type of dimension a value holds and what operations are possible with it. They are
+/// only used as template paramaters, are never instantiated and should only require forward declarations, but some
+/// compiler versions seem to fail if they aren't given real declarations.
+/// @{
+struct WidthTag{};
+struct HeightTag{};
+struct XTag{};
+struct YTag{};
+struct DeltaXTag{};
+struct DeltaYTag{};
+/// @}
+
+namespace detail
+{
+struct ValueWrapperBase{}; // Used for determining if a type is a wrapper
+}
+
+namespace generic
+{
+template<typename T>
+struct Value
+{
+    /// Wraps a geometry value and prevents it from being accidentally used for invalid operations (such as setting a
+    /// width to a height or adding two x positions together). Of course, explicit casts are possible to get around
+    /// these restrictions (see the as_*() functions). int values (which are most values) should use the
+    /// derived IntWrapper class, and other types should use this directly.
+    template<typename Tag>
+    struct Wrapper: detail::ValueWrapperBase
+    {
+        using ValueType = T;
+        using TagType = Tag;
+        template<typename OtherTag>
+        using WrapperType = Wrapper<OtherTag>;
```

Comment:
On reflection, I think the biggest problem I had following this was the name and usage of `WrapperType`. It is a metafunction for creating a _WholeValue_ type based on the same underlying value type and applying a supplied tag.

I would probably have written something using the TMP style, like:
```c++
template<typename OtherTag>
struct Corresponding { using type = Wrapper<OtherTag>; };
```
With the usage:
```c++
template<typename P, typename std::enable_if<std::is_base_of<detail::PointBase, P>::value, bool>::type = true>
inline constexpr P operator+(P lhs, typename P::template Corresponding<DeltaXTag>::type rhs) { return{lhs.x + rhs, lhs.y}; }
```
I'm not sure that my preference is enough for a "request changes", but `Corresponding<DeltaXTag>::type` makes more sense to me than "WrapperType<DeltaXTag>".

Reply:
That sounds like a better idea for the naming. I'll keep that in mind when I make another pass over this.